crypto/rand/randfile.c

```diff
@@ -83,26 +83,37 @@ int RAND_load_file(const char *file, long bytes)
         return 0;
 
 #ifndef OPENSSL_NO_POSIX_IO
-    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode)) {
-        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_NOT_A_REGULAR_FILE);
-        ERR_add_error_data(2, "Filename=", file);
-        return -1;
-    }
+    /*
+     * If not a regular file, assume it's a device and don't try to
+     * read everything.
+     */
+    if (stat(file, &sb) >= 0 && !S_ISREG(sb.st_mode) && bytes < 0)
+        bytes = 256 / 8;
 #endif
+
     if ((in = openssl_fopen(file, "rb")) == NULL) {
         RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_CANNOT_OPEN_FILE);
         ERR_add_error_data(2, "Filename=", file);
         return -1;
     }
 
+    /* Read the file, in chunks. */
     for ( ; ; ) {
         if (bytes > 0)
             n = (bytes < RAND_FILE_SIZE) ? (int)bytes : RAND_FILE_SIZE;
         else
             n = RAND_FILE_SIZE;
+
         i = fread(buf, 1, n, in);
```

Comment:
One has to recognize that fread a) buffers data and buffer is customarily larger than you ask for here (which spells waste and depletion, as already mentioned you have to use setvbuf); b) has internal loop that tries to fill at least as much as you asked for (which spells waiting for longer than intended without any chance to act upon it). As for b). Consider it's a pipe, do you actually want to wait? Forever? Well, one can always blame the application for using inadequate non-file, but one has to give it a chance to act, so maybe it shouldn't actually be a loop?

Reply:
On additional note setvbuf still loops till it fills the buffer...

Reply:
The suggestion wasn't to reduce the buffer but to remove it entirely: `setbuf(, NULL)`.
`setbuffer` could be used to buffer locally but it could block as you note.

For a regular file, this is moot.  We shouldn't adjust the buffer size and we should loop till completion.

It is only device node where there are problem cases -- we should try to read exactly what is asked, never a byte more and also to not wait forever.  This will be difficult using just stdio calls.  Two distinct paths seem to be required.

Reply:
yes, stdio with device files is a pain.